tools/revng-convert-ft/Main.cpp

```diff
@@ -0,0 +1,119 @@
+auto TaV = values(clABIDescription(SystemV_x86_64),
+                  clABIDescription(SystemV_x86),
+                  clABIDescription(SystemV_x86_regparm_3),
+                  clABIDescription(SystemV_x86_regparm_2),
+                  clABIDescription(SystemV_x86_regparm_1),
+                  clABIDescription(Microsoft_x86_64),
+                  clABIDescription(Microsoft_x86_64_vectorcall),
+                  clABIDescription(Microsoft_x86_64_clrcall),
+                  clABIDescription(Microsoft_x86_cdecl),
+                  clABIDescription(Microsoft_x86_stdcall),
+                  clABIDescription(Microsoft_x86_thiscall),
+                  clABIDescription(Microsoft_x86_fastcall),
+                  clABIDescription(Microsoft_x86_clrcall),
+                  clABIDescription(Microsoft_x86_vectorcall),
+                  clABIDescription(AArch64),
+                  clABIDescription(ARM),
+                  clABIDescription(SystemV_MIPS_o32),
+                  clABIDescription(SystemZ_s390x));
+opt<ABI> TargetABI("abi", Required, desc(TrgDesc), TaV, cat(Category));
+
+constexpr const char *FnDesc = "<input file name>";
+opt<std::string> Filename(Positional, Required, desc(FnDesc), cat(Category));
+
+opt<std::string> Output("o",
+                        desc("Optional output filename, if not specified, the "
+                             "output is dumped to `stdout`"),
+                        value_desc("path"),
+                        cat(Category));
+
+} // namespace Options
+
+int main(int argc, const char *argv[]) {
+  // Enable LLVM stack trace
+  llvm::sys::PrintStackTraceOnErrorSignal(argv[0]);
+
+  llvm::cl::HideUnrelatedOptions(Options::Category);
+  llvm::cl::ParseCommandLineOptions(argc, argv);
+
+  llvm::raw_fd_ostream *OutputStreamPtr;
+  if (Options::Output.empty()) {
+    OutputStreamPtr = &llvm::outs();
+  } else {
+    std::error_code EC;
+    static llvm::raw_fd_ostream OutputStream(Options::Output, EC);
+    if (!EC || OutputStream.has_error()) {
+      dbg << "Unable to open an output file: '" << Options::Output << "'.\n";
+      return 1;
+    }
+    OutputStreamPtr = &OutputStream;
+  }
+
+  auto BufferOrError = llvm::MemoryBuffer::getFileOrSTDIN(Options::Filename);
+  if (!BufferOrError) {
+    dbg << "Unable to open an input file: '" << Options::Filename << "'.\n";
+    return 2;
+  }
+  llvm::StringRef InputText = (*BufferOrError)->getBuffer();
+
+  auto Deserialized = TupleTree<model::Binary>::deserialize(InputText);
+  if (!Deserialized) {
+    dbg << "Unable to deserialize the model: '" << Options::Filename << "'.\n";
+    return 3;
+  }
+  if (!Deserialized->verify()) {
+    dbg << "Model verification failed: '" << Options::Filename << "'.\n";
+    return 4;
+  }
+
+  if (Options::Operation == Options::CABI)
+    return convertToCABI(*Deserialized, Options::TargetABI, *OutputStreamPtr);
+  else
+    return convertToRaw(*Deserialized, Options::TargetABI, *OutputStreamPtr);
```

Comment:
Let's see how long does it take me to implement `revng model opt`, but I'm pretty sure `revng-convert-ft` will become `revng model opt -convert-function-types-to-c-abi -target-abi=...` and `revng model opt -convert-to-raw-function-type`.
A first step in this direction would be to move `convertToCABI` and `convertToRaw` to `Processing.h`/`Processing.cpp`.

Reply:
https://github.com/revng/revng/pull/98/commits/a12741d8b1c9bf5a4127aac51f0d658efeac332c, what do you think?

Reply:
On that note, should I put the typo fix in its own commit or is it fine as is?

Reply:
Uhm, this is a new file so squash everything. Our policy on history is to make self-contained commits that fix/improve things that are already in `develop`. But for new things, the should look as if they were done right since the beginning, no fixup commits on newly introduced things.

Reply:
Sorry for misunderstanding, [this](https://github.com/revng/revng/commit/a12741d8b1c9bf5a4127aac51f0d658efeac332c#diff-1a81ec251d6b14c9ed05c4fc62dec55e21077ab2363912822820f556688de3c7R17) is the typo I'm referring to. It's in `Processing.h`.

I just asked it now since it's a part of the commit I mentioned here.

Reply:
If it's not too much of an hassle, make a commit at the beginning of the branch called "Minor changes" where you stuff all these things.

Reply:
Sure.

Reply:
Done and done.